### PR TITLE
Fix multiple annotation processing issues discovered by Micronaut

### PR DESCRIPTION
**Pull request by oehme**
- Don't reprocess types when there is nothing to compile
- Merge annotation processing data across subsequent compilations

Fixes https://github.com/micronaut-projects/micronaut-core/issues/6536

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by oehme** [comment]
Since this was a regression in 7.3, consider making this a 7.3.1 candidate.

**Comment by oehme** [comment]
@ljacomet FYI

**Comment by ljacomet** [comment]
Thanks for the fix Stefan!

We have plans for a 7.3.1, adding this.

**Comment by ljacomet** [comment]
Could you move these changes to be based off the `release` branch?

**Comment by oehme** [comment]
Done

**Comment by ljacomet** [comment]
@bot-gradle test this

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/AggregatingIncrementalAnnotationProcessingIntegrationTest.groovy
Comment (asodja) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/AggregatingIncrementalAnnotationProcessingIntegrationTest.groovy:439]:
```suggestion
    @Issue("https://github.com/micronaut-projects/micronaut-core/issues/6536")
```

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/IsolatingIncrementalAnnotationProcessingIntegrationTest.groovy
Comment (asodja) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/IsolatingIncrementalAnnotationProcessingIntegrationTest.groovy:64]:
```suggestion
    @Issue("https://github.com/micronaut-projects/micronaut-core/issues/6536")
```

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalResultStoringCompiler.java
Comment (asodja) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalResultStoringCompiler.java:181]: What is the reason we don't need to merge also other data here?

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java
Comment (asodja) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java:160]: Regarding the first point of comment:
```
This includes:
- types which are annotated with aggregating annotations, as aggregating processors need to see them regardless of what has changed
```
Now I believe this is not true anymore if there is nothing to compile. Is there any aggregating processor out there that needs to run every time, even if there is nothing to compile? I doubt that makes sense, but I would just want to understand the change better

**Review by asodja — COMMENTED** [review]
Looks good overall, I just added some questions to the change.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalResultStoringCompiler.java
Comment (oehme) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalResultStoringCompiler.java:181]: Aggregated types are always reprocessed, thus every compilation knows the full set of aggregated types.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java
Comment (oehme) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java:160]: "Regardless of what has changed" does not include "nothing has changed". I'm open for suggestion for better wording:) When nothing changed, processing makes no sense as you pointed out correctly.

**Review by asodja — APPROVED** [review]

**Comment by ljacomet** [comment]
@bot-gradle test and merge

**Comment by lightbody** [comment]
Hey - I'm eagerly waiting to try this out and see if it fixes our Micronaut build issues too. Unfortunately, I haven't figured out how to install Gradle locally, so my next best option is to wait for this to merge and be available in one of the nightly releases. 

Once merged, will it be in the 7.4 nightly or 7.3 nightly?

https://gradle.org/nightly/

https://gradle.org/release-nightly/

And of course, if you can point me to where I can solve my local Gradle install issues, I could verify this fix for our issues prior to merge ;)

**Comment by oehme** [comment]
@lightbody run `gradlew install` to create a local installation from source.

**Comment by lightbody** [comment]
@oehme Thanks. I was doing `gradle install` 😵‍💫 . 

I _think_ I'm using it now and, if so, it unfortunately doesn't solve my problem. We're seeing the same errors reported in that original Micronaut ticket. I'm going to try to triple check that I am indeed using the right build (I'm running `~/path/to/local/gradle compileJava` vs my normal `./gradlew compileJava`). But assuming it is indeed an outstanding bug, my next step will be to try to distill our giant build down to something that reproduces the problem.

**Comment by oehme** [comment]
@lightbody when you have a reproducer, just create a new issue and ping me on it and I'll look into it.

**Pull request merged**
